snapcraft/plugins/_python/_pip.py

```diff
@@ -0,0 +1,263 @@
+        # --ignore-installed: If the package is already installed, reinstall
+        #                     it. FIXME: Not sure why
+        self._run(['install', '--user', '--no-compile', '--no-index',
+                   '--find-links', self._python_package_dir,
+                   '--ignore-installed'] + list(args), **kwargs)
+
+    def wheel(self, args, **kwargs):
+        """Using the requested args, build wheels of packages in the cache.
+
+        Simple wrapper for `pip wheel`. Args are appended at the end, and
+        any keyword args are passed along to `common.run_output`. Note that
+        the packages should have already been downloaded via `self.download`.
+
+        :param list args: Arguments for `pip wheel`. Package names, for
+                          example, or the path to a dir containing a setup.py.
+
+        :return: List of paths to each wheel that was built.
+        :rtype: list
+        """
+        if not args:
+            return  # No wheels to build
+
+        wheels = []
+        with tempfile.TemporaryDirectory() as temp_dir:
+
+            # Using pip with a few special parameters:
+            #
+            # --no-index: Don't hit pypi, assume the packages are already
+            #             downloaded (i.e. by using `self.download()`)
+            # --find-links: Provide the directory into which the packages
+            #               should have already been fetched
+            # --wheel-dir: Build wheels into a temporary working area rather
+            #              rather than cwd. We'll copy them over. FIXME: Why
+            #              not just build them straight in the package_dir?
+            self._run(['wheel', '--no-index', '--find-links',
+                       self._python_package_dir, '--wheel-dir',
+                       temp_dir] + list(args), **kwargs)
+            wheels = os.listdir(temp_dir)
+            for wheel in wheels:
+                file_utils.link_or_copy(
+                    os.path.join(temp_dir, wheel),
+                    os.path.join(self._python_package_dir, wheel))
+
+        return [os.path.join(self._python_package_dir, wheel)
+                for wheel in wheels]
+
+    def list(self):
+        """Determine which packages have been installed.
+
+        :return: Dict of installed python packages and their versions
+        :rtype: dict
+        """
+        output = self._run(['list', '--format=json'])
+        packages = collections.OrderedDict()
+        try:
+            json_output = json.loads(
+                    output, object_pairs_hook=collections.OrderedDict)
+        except json.decoder.JSONDecodeError as e:
+            raise errors.PipListInvalidJsonError(output) from e
+
+        for package in json_output:
+            if 'name' not in package:
+                raise errors.PipListMissingFieldError('name', output)
+            if 'version' not in package:
+                raise errors.PipListMissingFieldError('version', output)
```

Comment:
Wait, is that possible? Shouldn't we just panic because pip is broken?

Reply:
Hey, we're calling an external command, anything is possible :stuck_out_tongue: . In all seriousness, no, I don't see this happening. But if it does, I think a bug report that says:

> I ran snapcraft and got "Pip packages json missing 'name' field"

Would be easier to triage than:

> I ran snapcraft and got this traceback:
>
> ```
> Traceback (most recent call last):
>   File "/snapcraft/bin/snapcraft", line 36, in <module>
>     obj=dict(project=snapcraft.ProjectOptions()))
>   File "/usr/lib/python3/dist-packages/click/core.py", line 716, in __call__
>     return self.main(*args, **kwargs)
>   File "/usr/lib/python3/dist-packages/click/core.py", line 696, in main
>     rv = self.invoke(ctx)
>   File "/usr/lib/python3/dist-packages/click/core.py", line 1037, in invoke
>     return Command.invoke(self, ctx)
>   File "/usr/lib/python3/dist-packages/click/core.py", line 889, in invoke
>     return ctx.invoke(self.callback, **ctx.params)
>   File "/usr/lib/python3/dist-packages/click/core.py", line 534, in invoke
>     return callback(*args, **kwargs)
>   File "/usr/lib/python3/dist-packages/click/decorators.py", line 17, in new_func
>     return f(get_current_context(), *args, **kwargs)
>   File "/snapcraft/snapcraft/cli/__init__.py", line 110, in run
>     ctx.forward(lifecyclecli.commands['snap'])
>   File "/usr/lib/python3/dist-packages/click/core.py", line 552, in forward
>     return self.invoke(cmd, **kwargs)
>   File "/usr/lib/python3/dist-packages/click/core.py", line 534, in invoke
>     return callback(*args, **kwargs)
>   File "/snapcraft/snapcraft/cli/lifecycle.py", line 135, in snap
>     project_options, directory=directory, output=output)
>   File "/snapcraft/snapcraft/internal/lifecycle.py", line 360, in snap
>     execute('prime', project_options)
>   File "/snapcraft/snapcraft/internal/lifecycle.py", line 130, in execute
>     _Executor(config, project_options).run(step, part_names)
>   File "/snapcraft/snapcraft/internal/lifecycle.py", line 225, in run
>     self._run_step(step, part, part_names)
>   File "/snapcraft/snapcraft/internal/lifecycle.py", line 262, in _run_step
>     getattr(part, step)()
>   File "/snapcraft/snapcraft/internal/pluginhandler/__init__.py", line 252, in pull
>     self.plugin.pull()
>   File "/snapcraft/snapcraft/plugins/python.py", line 180, in pull
>     self._run_pip(setup, download=True)
>   File "/snapcraft/snapcraft/plugins/python.py", line 340, in _run_pip
>     return pip.list(self.run_output)
>   File "/snapcraft/snapcraft/plugins/python.py", line 477, in list
>     packages[package['name']] = package['version']
> KeyError: 'name'
> ```

In fact, since the error is probably on their system, that helpful error might lead them to their problem, whereas the traceback is opaque. I guess my point is, it'd be nice to add some context to such errors if we have it. But indeed, this may be overly cautious. Shall I remove these exceptions?